File: lab02/LED.py
#!/usr/bin/python3

# script baseado no codigo disponibilizado em:
# Derek Molloy, Exploring Raspberry Pi: Interfacing to the Real World with Embedded Linux,
# Wiley 2016, ISBN 978-1-119-1868-1, http://www.exploringrpi.com/

import sys
from time import sleep

LED_NUMBER_VERMELHO = "20"
LED_PATH_VERMELHO = f"/sys/class/gpio/gpio{LED_NUMBER_VERMELHO}/"

LED_NUMBER_AMARELO = "16"
LED_PATH_AMARELO = f"/sys/class/gpio/gpio{LED_NUMBER_AMARELO}/"

LED_NUMBER_VERDE = "21"
LED_PATH_VERDE = f"/sys/class/gpio/gpio{LED_NUMBER_VERDE}/"

SYSFS_DIR = "/sys/class/gpio/"

contador = 0

def writeLED ( filename, value, path ):
        "Esta funcao escreve o valor 'value' no arquivo 'path+filename'"
        fo = open( path + filename,"w")
        fo.write(value)
        fo.close()
        return

print("Habilitando a gpio")
#HABILITA GPIO VERMELHO
writeLED (filename="export", value=LED_NUMBER_VERMELHO, path=SYSFS_DIR)
writeLED (filename="export", value=LED_NUMBER_AMARELO, path=SYSFS_DIR)
writeLED (filename="export", value=LED_NUMBER_VERDE, path=SYSFS_DIR)
sleep(0.1)
writeLED (filename="direction", value="out", path = LED_PATH_VERMELHO)
writeLED (filename="direction", value="out", path = LED_PATH_AMARELO)
writeLED (filename="direction", value="out", path = LED_PATH_VERDE)

while contador <= 5:
	# LIGA LED VERMELHO
	writeLED (filename=f"value", value="1", path = LED_PATH_VERMELHO)
	sleep(2)
	writeLED (filename=f"value", value="0", path = LED_PATH_VERMELHO)

#
	# LIGA LED AMARELO
	writeLED (filename=f"value", value="1", path = LED_PATH_AMARELO)
	sleep(1)
	writeLED (filename=f"value", value="0", path = LED_PATH_AMARELO)

	# LIGA LED VERDE
	writeLED (filename=f"value", value="1", path = LED_PATH_VERDE)
	sleep(1)
	writeLED (filename=f"value", value="0", path = LED_PATH_VERDE)

	contador += 1

writeLED (filename="unexport", value=LED_NUMBER_VERMELHO, path=SYSFS_DIR)
writeLED (filename="unexport", value=LED_NUMBER_AMARELO, path=SYSFS_DIR)
writeLED (filename="unexport", value=LED_NUMBER_VERDE, path=SYSFS_DIR)

File: lab02/LED.cpp
// Barbara Tong
#include <iostream>
#include <fstream>
#include <string>
#include <unistd.h>
using namespace std;

// define variaveis para os leds
#define GPIO_AMARELO "16"
#define GPIO_PATH_AMARELO "/sys/class/gpio/gpio16/"
#define GPIO_SYSFS_AMARELO "/sys/class/gpio/"

#define GPIO_VERMELHO "20"
#define GPIO_PATH_VERMELHO "/sys/class/gpio/gpio20/"
#define GPIO_SYSFS_VERMELHO "/sys/class/gpio/"

#define GPIO_VERDE "21"
#define GPIO_PATH_VERDE "/sys/class/gpio/gpio21/"
#define GPIO_SYSFS_VERDE "/sys/class/gpio/"

// define funcao que escreve 0 ou 1 para ligar o led
void writeGPIO(string path, string filename, string value)
{
    fstream fs;
    fs.open((path + filename).c_str(), fstream::out);
    fs << value;
    fs.close();
}

int main(int argc, char *argv[])
{
    // contador
    int counter = 0

    cout << "Habilitando a gpio" << endl;
    writeGPIO(string(GPIO_SYSFS_AMARELO), "export", GPIO_AMARELO);
    usleep(100000);
    writeGPIO(string(GPIO_PATH_AMARELO), "direction", "out");

    writeGPIO(string(GPIO_SYSFS_VERMELHO), "export", GPIO_VERMELHO);
    usleep(100000);
    writeGPIO(string(GPIO_PATH_VERMELHO), "direction", "out");

    writeGPIO(string(GPIO_SYSFS_VERDE), "export", GPIO_VERDE);
    usleep(100000);
    writeGPIO(string(GPIO_PATH_VERDE), "direction", "out");

    while (counter <= 5)
    {
        cout << "Acendendo o LED Vermelho" << endl;
        writeGPIO(string(GPIO_PATH_VERMELHO), "value", "1");
        sleep(2)
        writeGPIO(string(GPIO_PATH_VERMELHO), "value", "0");

        cout << "Acendendo o LED Amarelo" << endl;
        writeGPIO(string(GPIO_PATH_AMARELO), "value", "1");
        sleep(1)
        writeGPIO(string(GPIO_PATH_AMARELO), "value", "0");

        cout << "Acendendo o LED Verde" << endl;
        writeGPIO(string(GPIO_PATH_VERDE), "value", "1");
        sleep(1)
        writeGPIO(string(GPIO_PATH_VERDE), "value", "0");

        counter++
    }

    writeGPIO(string(GPIO_SYSFS_AMARELO), "unexport", GPIO_AMARELO);
    writeGPIO(string(GPIO_SYSFS_VERMELHO), "unexport", GPIO_VERMELHO);
    writeGPIO(string(GPIO_SYSFS_VERDE), "unexport", GPIO_VERDE);
    
    return 0;
}